File: backend/app/schemas/schedule.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field


# ==========================================
# Schedule Entry
# ==========================================
class ScheduleEntryCreate(BaseModel):
    class_id: UUID
    section: str | None = None
    subject_id: UUID | None = None
    subject_name: str
    teacher_id: UUID | None = None
    teacher_name: str | None = None
    room_id: UUID | None = None
    room_name: str | None = None
    day: str
    period: int = Field(ge=1, le=20)


class ScheduleEntryUpdate(BaseModel):
    class_id: UUID | None = None
    section: str | None = None
    subject_id: UUID | None = None
    subject_name: str | None = None
    teacher_id: UUID | None = None
    teacher_name: str | None = None
    room_id: UUID | None = None
    room_name: str | None = None
    day: str | None = None
    period: int | None = Field(default=None, ge=1, le=20)


class ScheduleEntryResponse(BaseModel):
    id: UUID | None = None
    institution_id: UUID
    class_id: UUID
    class_name: str | None = None
    section: str | None = None
    subject_id: UUID | None = None
    subject_name: str
    teacher_id: UUID | None = None
    teacher_name: str | None = None
    room_id: UUID | None = None
    room_name: str | None = None
    day: str
    period: int
    source: str = "manual"
    created_at: datetime | None = None
    updated_at: datetime | None = None


# ==========================================
# Timetable Generation
# ==========================================
class ScheduleGenerateRequest(BaseModel):
    class_ids: list[UUID] | None = Field(
        default=None,
        description="Classes to schedule. Omit to schedule every class in the institution.",
    )
    working_days: list[str] = Field(
        default=["Monday", "Tuesday", "Wednesday", "Thursday", "Friday"]
    )
    periods_per_day: int = Field(default=6, ge=1, le=12)
    sessions_per_week: int = Field(default=1, ge=1, le=10)
    persist: bool = Field(
        default=True,
        description="Persist generated slots to schedule_entries (replaces previously generated slots).",
    )


class ScheduleGenerateResponse(BaseModel):
    status: str
    message: str
    generated: list[ScheduleEntryResponse]
    skipped: list[dict]
    conflicts: int = 0


# ==========================================
# Conflict Detection
# ==========================================
class ScheduleConflict(BaseModel):
    type: str
    day: str
    period: int
    value: str
    entries: list[UUID]


class ScheduleConflictResponse(BaseModel):
    institution_id: UUID
    conflicts: list[ScheduleConflict]


# ==========================================
# Lookup data for the frontend
# ==========================================
class ScheduleClass(BaseModel):
    id: UUID
    name: str
    section: str | None = None


class ScheduleTeacher(BaseModel):
    id: UUID
    name: str | None = None


class ScheduleSubject(BaseModel):
    id: UUID
    name: str | None = None

File: frontend/src/lib/services/schedule.ts
import { API } from '$lib/config/api';
import { apiFetch } from '$lib/services/apiClient';

export interface ScheduleClass {
	id: string;
	name: string;
	section?: string | null;
}

export interface ScheduleTeacher {
	id: string;
	name?: string | null;
}

export interface ScheduleSubject {
	id: string;
	name?: string | null;
}

export interface ScheduleEntry {
	id?: string | null;
	institution_id: string;
	class_id: string;
	class_name?: string | null;
	section?: string | null;
	subject_id?: string | null;
	subject_name: string;
	teacher_id?: string | null;
	teacher_name?: string | null;
	room_id?: string | null;
	room_name?: string | null;
	day: string;
	period: number;
	source?: string;
}

export interface GenerateRequest {
	class_ids?: string[] | null;
	working_days: string[];
	periods_per_day: number;
	sessions_per_week: number;
	persist: boolean;
}

export interface GenerateResult {
	status: string;
	message: string;
	generated: ScheduleEntry[];
	skipped: { class?: string; subject?: string | null; reason?: string }[];
	conflicts: number;
}

export interface ScheduleConflict {
	type: string;
	day: string;
	period: number;
	value: string;
	entries: string[];
}

const base = `${API.baseUrl}/api/schedule`;

export async function getScheduleClasses(): Promise<ScheduleClass[]> {
	return apiFetch<ScheduleClass[]>(`${base}/classes`);
}

export async function getScheduleTeachers(): Promise<ScheduleTeacher[]> {
	return apiFetch<ScheduleTeacher[]>(`${base}/teachers`);
}

export async function getScheduleSubjects(): Promise<ScheduleSubject[]> {
	return apiFetch<ScheduleSubject[]>(`${base}/subjects`);
}

export async function getTimetable(
	options: { class_id?: string; teacher_id?: string; day?: string } = {}
): Promise<ScheduleEntry[]> {
	const params = new URLSearchParams();
	if (options.class_id) params.set('class_id', options.class_id);
	if (options.teacher_id) params.set('teacher_id', options.teacher_id);
	if (options.day) params.set('day', options.day);
	const qs = params.toString();
	return apiFetch<ScheduleEntry[]>(`${base}${qs ? `?${qs}` : ''}`);
}

export async function getTimetableForClass(classId: string): Promise<ScheduleEntry[]> {
	return apiFetch<ScheduleEntry[]>(`${base}/class/${classId}`);
}

export async function getTimetableForTeacher(teacherId: string): Promise<ScheduleEntry[]> {
	return apiFetch<ScheduleEntry[]>(`${base}/teacher/${teacherId}`);
}

export async function getScheduleConflicts(): Promise<ScheduleConflict[]> {
	const result = await apiFetch<{ institution_id: string; conflicts: ScheduleConflict[] }>(
		`${base}/conflicts`
	);
	return result.conflicts;
}

export async function generateTimetable(request: GenerateRequest): Promise<GenerateResult> {
	return apiFetch<GenerateResult>(`${base}/generate`, {
		method: 'POST',
		body: JSON.stringify(request)
	});
}

export async function deleteScheduleEntry(entryId: string): Promise<void> {
	await apiFetch(`${base}/${entryId}`, { method: 'DELETE' });
}